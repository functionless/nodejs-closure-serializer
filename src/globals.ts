import { parseSync } from "@swc/core";

/**
 * This file adds the `__fnl_func` function to the global object.
 *
 * It must be run prior to the `@swc/register` hook runs as part of the `node -r <require>` step or else
 * transformed code will break.
 */

export interface UnparsedClosure {
  filename: string;
  captured: () => CapturedVariables;
}

type CapturedVariables = any[];

export interface Closure {
  filename: string;
  captured: Record<string, any>;
}

export type BoundFunction = [
  targetFunc: Function,
  boundThis: any,
  boundArgs: any[]
];

declare global {
  function __fnl_func(
    func: Function,
    filename: string,
    captured: () => CapturedVariables
  ): void;
}
global.__fnl_func = registerClosure;

export function registerClosure(
  func: Function,
  filename: string,
  captured: () => CapturedVariables
) {
  if (closures.has(func)) {
    throw new Error(`illegal override of function captured closure`);
  }
  closures.set(func, {
    filename,
    captured: captured,
  });
  return func;
}

const closures = globalSingleton(
  Symbol.for("@functionless/captured-closures"),
  () => new WeakMap<Function, UnparsedClosure>()
);

const bind = Function.prototype.bind;

Function.prototype.bind = function (boundThis: any, ...boundArgs: any[]) {
  const bound = bind.call(this, boundThis, ...boundArgs);
  boundFunctions.set(bound, [this, boundThis, boundArgs]);
  return bound;
};

const boundFunctions = globalSingleton(
  Symbol.for("@functionless/bound-functions"),
  () => new WeakMap<Function, BoundFunction>()
);

// cache parsed closures
const parsedClosureCache = globalSingleton(
  Symbol.for("@functionless/parsed-closure-cache"),
  () => new WeakMap<Function, Closure>()
);

export function getClosure(func: Function): Closure | undefined {
  const closure = closures.get(func);
  if (closure === undefined) {
    return undefined;
  }

  if (!parsedClosureCache.has(func)) {
    // because of SWC's hygiene step where variables are renamed
    // we cannot inject the name of a variable as a string literal
    // as a workaround, we parse the closure that returns the free variables
    // which is in the form: () => [a, b, c]

    const ast = parseSync(closure.captured.toString(), {
      syntax: "ecmascript",
      script: false,
    });

    if (ast.body.length !== 1) {
      throw new Error(`expected a single statement`);
    }
    const stmt = ast.body[0]!;
    if (stmt.type !== "ExpressionStatement") {
      throw new Error(`expected an ExpressionStatement, but got ${stmt.type}`);
    } else if (stmt.expression.type !== "ArrowFunctionExpression") {
      throw new Error(
        `expected ExpressionStatement's expression to be an ArrowFunctionExpression, but got ${stmt.expression.type}`
      );
    } else if (stmt.expression.body.type !== "ArrayExpression") {
      throw new Error(
        `expected ArrowFunctionExpressions's body to be an ArrayExpression, but got ${stmt.expression.body.type}`
      );
    }

    const variableValues = closure.captured();

    if (stmt.expression.body.elements.length !== variableValues.length) {
      throw new Error(
        `expected ArrayExpression to have ${variableValues.length} elements, but has ${stmt.expression.body.elements.length}`
      );
    }

    parsedClosureCache.set(func, {
      filename: closure.filename,
      captured: Object.fromEntries(
        stmt.expression.body.elements.map((variable, i) => {
          if (variable?.expression.type !== "Identifier") {
            throw new Error(
              `expected item ${i} in ArrayExpression to be an Identifier, but got ${variable?.expression.type}`
            );
          }

          return [variable.expression.value, variableValues[i]];
        })
      ),
    });
  }

  return parsedClosureCache.get(func);
}

export function getBoundFunction(func: Function): BoundFunction | undefined {
  return boundFunctions.get(func);
}

function globalSingleton<T>(key: symbol, create: () => T): T {
  return ((global as any)[key] = (global as any)[key] ?? create());
}
