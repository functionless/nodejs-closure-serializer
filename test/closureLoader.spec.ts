// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* eslint-disable */

// Make sure we are listening for v8 events as they're necessary to get things like file locations
// for serialization errors.  We need to do this first, before we even get around to running tests.
import * as v8Hooks from "../src/closure/v8Hooks";
import { describe, test, beforeAll, jest, expect } from "@jest/globals";

jest.setTimeout(10000);

// only start running these tests once we've initialized the code in v8Hooks.
describe("after hooks", () => {
    let done = false;
    beforeAll(async  () => {
        await v8Hooks.isInitializedAsync();
        done = true;
    });

    require("./tsClosureCases");

    test("invoke", () => {
        expect(done).toBeTruthy();
    })
});