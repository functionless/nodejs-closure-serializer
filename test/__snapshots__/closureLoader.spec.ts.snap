// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`after hooks closure Analyze property chain #9 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

var __o = {};
var __o_b = {};
Object.defineProperty(__o_b, \\"name\\", { configurable: true, value: \\"b\\" });
__o.b = __o_b;

function __f0() {
  return (function() {
    with({ o: __o }) {

return function () { console.log(o.b.name); };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;

exports[`after hooks closure Class complex statics 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

__f1.S = __f2;
__f1.SX = __f3;
__f1.x = 3;

function __f1() {
  return (function() {
    with({  }) {

return function /*constructor*/() { };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f2() {
  return (function() {
    with({  }) {

return () => \\"x\\";

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f3() {
  return (function() {
    with({ ClassWithMoreStatic: __f1 }) {

return () => new ClassWithMoreStatic();

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f0() {
  return (function() {
    with({ ClassWithMoreStatic: __f1 }) {

return () => {
            return new ClassWithMoreStatic();
        };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;

exports[`after hooks closure Class complex statics only 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

__f1.S = __f2;
__f1.SX = __f3;
__f1.x = 3;

function __f1() {
  return (function() {
    with({  }) {

return function /*constructor*/() { };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f2() {
  return (function() {
    with({  }) {

return () => \\"x\\";

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f3() {
  return (function() {
    with({ ClassWithMoreStatic: __f1 }) {

return () => new ClassWithMoreStatic();

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f0() {
  return (function() {
    with({ ClassWithMoreStatic: __f1 }) {

return () => {
            ClassWithMoreStatic.S();
            ClassWithMoreStatic.SX();
            console.log(ClassWithMoreStatic.x);
        };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;

exports[`after hooks closure Class nested statics 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

var __c = {};
var __c_x = {S: \\"hello\\", T: \\"hello2\\"};
__c.x = __c_x;

function __f0() {
  return (function() {
    with({ c: __c }) {

return () => {
            return c.x.S + c.x.T;
        };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;

exports[`after hooks closure Class statics 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

var __ClassWithStatic = {S: \\"hello\\"};

function __f0() {
  return (function() {
    with({ ClassWithStatic: __ClassWithStatic }) {

return () => {
            return ClassWithStatic.S;
        };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;

exports[`after hooks closure Class statics with class invoke 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

__f1.S = \\"hello\\";
__f1.T = \\"hello2\\";

function __f1() {
  return (function() {
    with({  }) {

return function /*constructor*/() { };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f0() {
  return (function() {
    with({ ClassWithStatic: __f1 }) {

return () => {
            const s = ClassWithStatic.S;
            const a = new ClassWithStatic();
            return s;
        };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;

exports[`after hooks closure Class statics with prototype reference 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

__f1.S = \\"hello\\";
__f1.T = \\"hello2\\";

function __f1() {
  return (function() {
    with({  }) {

return function /*constructor*/() { };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f0() {
  return (function() {
    with({ ClassWithStatic: __f1 }) {

return () => {
            console.log(ClassWithStatic);
            return ClassWithStatic.S;
        };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;

exports[`after hooks closure Class statics without static reference 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

__f1.S = \\"hello\\";
__f1.T = \\"hello2\\";

function __f1() {
  return (function() {
    with({  }) {

return function /*constructor*/() { };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f0() {
  return (function() {
    with({ ClassWithStatic: __f1 }) {

return () => {
            return new ClassWithStatic();
        };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;

exports[`after hooks closure Function invoked with props 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

__F.x = \\"hello\\";

function __F() {
  return (function() {
    with({ F: __F }) {

return function /*F*/() {
        return \\"wee\\";
    };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}

function __f0() {
  return (function() {
    with({ F: __F }) {

return () => {
            console.log(F());
        };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;

exports[`after hooks closure Function with props 1`] = `
Object {
  "exportName": "handler",
  "text": "exports.handler = __f0;

var __F = {x: \\"hello\\"};

function __f0() {
  return (function() {
    with({ F: __F }) {

return () => {
            console.log(F.x);
        };

    }
  }).apply(undefined, undefined).apply(this, arguments);
}
",
}
`;
